File: src/main/java/org/willxu/algorithm/service/bool/PalindromeNumber.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.bool;

public interface PalindromeNumber {
    /**
     * Given an integer x, return true if x is palindrome integer.
     * An integer is a palindrome when it reads the same backward as
     * forward.
     *
     * For example, 121 is palindrome while 123 is not.
     *
     * @param x -2^31 <= x <= 2^31 - 1
     * @return check palindrome integer
     */
    boolean isPalindrome(int x);
}

File: src/leetcode/p0000_0099/p0009_palindrome_number/palindrome_number.py
from abc import ABC, abstractmethod
from typing import List, Dict

class PalindromeNumber(ABC):
    """
    Given an integer x, return true if x is a palindrome, and false otherwise.

    An integer is a palindrome when it reads the same forward and backward.
    For example, 121 is a palindrome while 123 is not.
    """
    @abstractmethod
    def is_palindrome(self, x: int) -> bool:
        """
        Args:
            -2^31 <= x <= 2^31 - 1
        """
        pass

class PalindromeNumberArray(PalindromeNumber):
    def is_palindrome(self, x: int) -> bool:
        if x == 0:
            return True;
        if x < 0 or x % 10 == 0:
            return False
        reversed_num = 0
        while x > reversed_num:
            reversed_num = reversed_num * 10 + x % 10
            x //= 10
        return x == reversed_num or x == reversed_num // 10
